marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/entity-validation-lib.sjs

```diff
@@ -0,0 +1,223 @@
+    if ("xml" == options.outputFormat) {
+      validateXmlEntity(newInstance, options, value);
+    } else {
+      validateJsonEntity(newInstance, options, value, entityInfo);
+    }
+  }
+}
+
+function shouldValidateEntity(options) {
+  let value = options.validateEntity;
+  if (value != null && value != undefined) {
+    value = fn.string(value).toLowerCase();
+    return value == "accept" || value == "reject";
+  }
+  return false;
+}
+
+function validateJsonEntity(newInstance, options, validateEntityValue, entityInfo) {
+  // As of 5.1.0, this is safe to do. But eventually, we'll want to find a schema by querying the schema database, or,
+  // better yet, via some API function call that does the work for us.
+  const entitySchemaUri = "/entities/" + entityInfo.title + ".entity.schema.json";
+  try {
+    xdmp.jsonValidate(newInstance, entitySchemaUri, ["full"]);
+  } catch (e) {
+    if ("accept" == validateEntityValue) {
+      if (options.headers == null) {
+        options.headers = {};
+      }
+      if (options.headers.datahub == null) {
+        options.headers.datahub = {};
+      }
+
+      // Tossing information about the errors in the headers so that they're added to the envelope.
+      // Note that regardless of the number of errors, xdmp.jsonValidate will return all of them in a "data" array, with
+      // each array item being a string.
+      options.headers.datahub.validationErrors = {
+        "name": e.name,
+        "data": e.data,
+        "message": e.message,
+        "formattedMessages": []
+      }
+
+      for (let errorMessage of e.data) {
+        let formattedErrorMessage = formatErrorMessageForJson(errorMessage.toString());
+        if (formattedErrorMessage == null) {
+          formattedErrorMessage = {
+            message : errorMessage.trim()
+          };
+        }
+        options.headers.datahub.validationErrors.formattedMessages.push(formattedErrorMessage);
+      }
+    } else if ("reject" == validateEntityValue) {
+      throw Error(e);
+    }
+  }
+}
+
+/**
+ * Attempts to format the given error message into a JSON object with a propertyName and a message that is friendlier to
+ * a human. Only supports "Missing property: Required" messages so far.
+ */
+function formatErrorMessageForJson(errorMessage) {
+  const missingMessage = "Missing property: Required";
+  let pos = errorMessage.indexOf(missingMessage);
+  let formattedErrorMessage = null;
+  if (pos > -1) {
+    errorMessage = errorMessage.substring(pos + missingMessage.length).trim();
+    pos = errorMessage.indexOf("not found");
+    if (pos > -1) {
+      errorMessage = errorMessage.substring(0, pos + "not found".length);
+      let propertyName = errorMessage.split(" ")[0];
+      formattedErrorMessage = {
+        propertyName: propertyName,
+        message: "Required " + errorMessage
+      };
+    }
+  }
+  return formattedErrorMessage;
+}
+
+function validateXmlEntity(newInstance, options, validateEntityValue) {
```

Comment:
options. null check please

Reply:
options is guaranteed to not be null here. "validateEntity" is the entry point to the library, I'll add a null-check there - though FWIW, the only client of validateEntity is entity-services/main.sjs, and options is guaranteed to exist there as well.

Reply:
Check the scope within the property it is defined in.

in this case, the function block below.